File: android/src/main/java/com/sentinelsdk/AuthModule.kt
package com.sentinelsdk

import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.Callback
import com.facebook.react.bridge.Promise
import com.qxdev.sentinel_sdk.SentinelProvider
import com.qxdev.sentinel_sdk.cloud.data.Response
import com.qxdev.sentinel_sdk.cloud.data.auth.User
import com.qxdev.sentinel_sdk.cloud.data.auth.UserData
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch

import com.facebook.react.bridge.WritableMap
import com.facebook.react.bridge.Arguments

import com.facebook.react.modules.core.DeviceEventManagerModule

class AuthModule(reactContext: ReactApplicationContext) :
  ReactContextBaseJavaModule(reactContext) {

  override fun getName(): String {
    return NAME
  }

  private val coroutineScope = CoroutineScope(Dispatchers.IO + Job())
  private val cloudProvider = SentinelProvider.cloudProvider
  private val sessionProvider = SentinelProvider.sessionProvider
  private val authServices = cloudProvider.authCloudServices

  // Example method
  // See https://reactnative.dev/docs/native-modules-android

  override fun initialize() {
    super.initialize()
    coroutineScope.launch {
      sessionProvider.hasActiveSession.collect {
        val event = Arguments.createMap()
        event.putBoolean("localSession", sessionProvider.hasActiveSession.value)
        sendEvent("hasSession", event)
      }
    }
  }

  @ReactMethod
  fun isLoggedIn(promise: Promise) {
    promise.resolve(sessionProvider.hasActiveSession.value)
  }

  private fun sendEvent(eventName: String, params: WritableMap) {
    reactApplicationContext
      .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
      .emit(eventName, params)
  }

  @ReactMethod
  fun signIn(username: String, password: String, promise: Promise) {
    coroutineScope.launch {
      val resultado = kotlin.runCatching {
        authServices.login(User(username.trim(), password.trim()))
      }
      val result = resultado.getOrNull()

      if (result != null) {
        if (result.success) {
          promise.resolve("${result.data}")
        } else {
          promise.reject("${result.errorCode}", "${result.errorCode}")
        }
      } else {
        println("Request failed with Exception ${resultado.exceptionOrNull()?.message}")
        promise.reject("${0}", "0")
      }
    }
  }

  @ReactMethod
  fun signIn(
    username: String,
    password: String,
    failureCallback: Callback,
    successCallback: Callback
  ) {
    coroutineScope.launch {
      val result: Response<Any> = authServices.login(User(username.trim(), password.trim()))
      if (result.success) {
        successCallback.invoke("${result.data}")
      } else {
        failureCallback.invoke("${result.errorCode}", "${result.errorCode}")
      }
    }
  }

  @ReactMethod
  fun recoverPassword(email: String, promise: Promise) {
    coroutineScope.launch {
      val result = authServices.changePassword(UserData(email.trim()))
      try {
        if (result.success) {
          promise.resolve("${result.data}")
        } else {
          promise.reject("${result.errorCode}", "${result.errorCode}")
        }
      } catch (error: Error) {
        promise.reject("Error", error)
      }
    }
  }

  @ReactMethod
  fun recoverPassword(email: String, failureCallback: Callback, successCallback: Callback) {
    coroutineScope.launch {
      val result = authServices.changePassword(UserData(email.trim()))
      if (result.success) {
        successCallback.invoke("${result.data}")
      } else {
        failureCallback.invoke("${result.errorCode}", "${result.errorCode}")
      }
    }
  }

  @ReactMethod
  fun signOut(promise: Promise) {
    coroutineScope.launch {
      try {
        val result = authServices.logout()
        if (result.success) {
          promise.resolve("${result.data}")
        } else {
          promise.reject("${result.errorCode}", "${result.errorCode}")
        }
      } catch (e: Exception) {
        sessionProvider.sessionData = null
        promise.reject("Error", e)
      }
    }
  }

  @ReactMethod
  fun signOut(failureCallback: Callback, successCallback: Callback) {
    coroutineScope.launch {
      val result = authServices.logout()
      if (result.success) {
        successCallback.invoke("${result.data}")
      } else {
        failureCallback.invoke("${result.errorCode}", "${result.errorCode}")
      }
    }
  }

  @ReactMethod
  fun accountType(promise: Promise) {
    coroutineScope.launch {
      val resultado = kotlin.runCatching {
        authServices.getAccountType()
      }
      val result = resultado.getOrNull()
      if (result != null) {
        promise.resolve("$result")
      } else {
        promise.reject("$result", "$result")
      }
    }
  }

  companion object {
    const val NAME = "AuthModule"
  }
}

File: ios/AuthModule.swift
import Foundation
import shared

@objc(AuthModule)
class AuthModule: RCTEventEmitter {
    
    private let cloudProvider = SentinelProvider.shared.cloudProvider
    private let sessionProvider = SentinelProvider.shared.sessionProvider
    private var authServices: AuthServices
    
    override func supportedEvents() -> [String]! {
        return ["hasSession"]
    }
    
    @objc
    override init () {
        self.authServices = cloudProvider.authCloudServices
        super.init()
        // TODO: comment this back in when we have access to the shared library.
//        sessionProvider.hasActiveSession.collect {
//            var event = Arguments.createMap()
//            event.putBoolean("localSession", sessionProvider.hasActiveSession.value)
//            sendEvent("hasSession", event)
//        }
    }
    
    @objc
    func isLoggedIn(_ resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
        resolve(sessionProvider.hasActiveSession.value)
    }
        
    @objc
    func signIn(_ username: String, password: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        Task { @MainActor in
            do {
                print("signIn::Promise")
                let loginResult = try await authServices.login(user: User(email: username.trimmingCharacters(in: .whitespacesAndNewlines), password: password.trimmingCharacters(in: .whitespacesAndNewlines)))
                
                if (loginResult.success) {
                    resolve("${result.data}")
                } else {
                    let error = NSError(domain: "", code: loginResult.errorCode as? Int ?? 0, userInfo: ["${result.errorCode}": "${result.errorCode}"])
                    reject("ERROR_SIGNIN", "There was an error sign in", error)
                }

            }
            catch {
                print("Request failed with Exception ${error.localizedDescription}")
                print(error)
                let error = NSError(domain: "", code: 0, userInfo: ["${0}": error.localizedDescription])
                reject("ERROR_SIGNIN", "Exception", error)
            }
        }
    }
    
    @objc
    func signIn(_ username: String, password: String, failureCallback: @escaping RCTResponseSenderBlock, successCallback: @escaping RCTResponseSenderBlock) {
        print("signIn::Callback")
        Task {
            do {
                let result = try await authServices.login(user: User(email: username.trimmingCharacters(in: .whitespacesAndNewlines), password: password.trimmingCharacters(in: .whitespacesAndNewlines)))
                
                if (result.success) {
                    successCallback(["${result.data}"])
                } else {
                    failureCallback(["${result.errorCode}", "${result.errorCode}"])
                }
            } catch {
                print(error)
                let error = NSError(domain: "", code: 0, userInfo: ["${0}": error.localizedDescription])
                failureCallback(["ERROR_SIGNIN", "Exception", error])
            }
        }
    }
    
    @objc
    func recoverPassword(_ email: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        Task { @MainActor in
            do {
                let changePasswordResult = try await authServices.changePassword(email: UserData(email: email.trimmingCharacters(in: .whitespacesAndNewlines)))
                
                if (changePasswordResult.success) {
                    resolve("${changePasswordResult.data}")
                } else {
                    let error = NSError(domain: "", code: changePasswordResult.errorCode as? Int ?? 0, userInfo: ["${changePasswordResult.errorCode}": "${changePasswordResult.errorCode}"])
                    reject("ERROR_RECOVERPASSWORD", "There was an error recovering the password", error)
                }
            } catch {
                let error = NSError(domain: "", code: 0, userInfo: ["Error": "Exception"])
                reject("ERROR_RECOVERPASSWORD", "An exception happen while recovering the password", error)
            }
        }
    }
    
    @objc
    func recoverPassword(_ email: String, failureCallback: @escaping RCTResponseSenderBlock, successCallback: @escaping RCTResponseSenderBlock) {
        Task { @MainActor in
            do {
                let  changePasswordResult = try await authServices.changePassword(email: UserData(email: email.trimmingCharacters(in: .whitespacesAndNewlines)))
                if (changePasswordResult.success) {
                    successCallback(["${changePasswordResult.data}"])
                } else {
                    failureCallback(["${changePasswordResult.errorCode}", "${changePasswordResult.errorCode}"])
                }
            } catch {
                let error = NSError(domain: "", code: 0, userInfo: ["${0}": error.localizedDescription])
                failureCallback(["ERROR_SIGNIN", "Exception", error])
            }

        }
    }
    
    @objc
    func signOut(_ resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        Task { @MainActor in
            do {
                let signOutResult = try await authServices.logout()
                
                if (signOutResult.success) {
                    resolve("${signOutResult.data}")
                } else {
                    let error = NSError(domain: "", code: signOutResult.errorCode as! Int, userInfo: ["${signOutResult.errorCode}": "${signOutResult.errorCode}"])
                    reject("ERROR_LOGINOUT", "There was an error login out", error)
                }
            }
            catch {
                print("Request failed with Exception ${error.localizedDescription}")
                print(error)
                let error = NSError(domain: "", code: 0, userInfo: ["${0}": error.localizedDescription])
                reject("ERROR_LOGINOUT", "Exception", error)
            }
        }
    }
    
    @objc
    func signOut(_ failureCallback: @escaping RCTResponseSenderBlock, successCallback: @escaping RCTResponseSenderBlock) {
        Task { @MainActor in
            do {
                let signOutResult = try await authServices.logout()
                if (signOutResult.success) {
                    successCallback(["${signOutResult.data}"])
                } else {
                    failureCallback(["${signOutResult.errorCode}", "${result.errorCode}"])
                }
            } catch {
                let error = NSError(domain: "", code: 0, userInfo: ["${0}": error.localizedDescription])
                failureCallback(["ERROR_SIGNIN", "Exception", error])
            }
        }
    }
    
    @objc
    func accountType(_ resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
        
        Task { @MainActor in
            do {
                var accountTypeResult = try await authServices.getAccountType()
                if (accountTypeResult != "") {
                    resolve(accountTypeResult)
                } else {
                    let error = NSError(domain: "", code: 0, userInfo: ["$accountTypeResult": "$accountTypeResult"])
                    reject("ERROR_ACCOUNTTYPE", "An error happen while getting the account type", error)
                }
            } catch {
                let error = NSError(domain: "", code: 0, userInfo: ["${0}": error.localizedDescription])
                reject("ERROR_ACCOUNTYPE", "Exception", error)
            }
        }
    }
    
    override static func requiresMainQueueSetup() -> Bool {
        return true;
    }
}
